src/wallet/wallet2.cpp

```diff
@@ -5437,23 +5779,67 @@ void wallet2::get_outs(std::vector<std::vector<tools::wallet2::get_outs_entry>>
+      if (td.m_key_image_known && !td.m_key_image_partial)
+      {
+        std::vector<uint64_t> ring;
+        if (get_ring(key, td.m_key_image, ring))
+        {
+          MINFO("This output has a known ring, reusing (size " << ring.size() << ")");
+          THROW_WALLET_EXCEPTION_IF(ring.size() > fake_outputs_count + 1, error::wallet_internal_error,
+              "An output in this transaction was previously spent on another chain with ring " +
+              std::to_string(ring.size()) + ", it cannot be spent now with ring size " +
+              std::to_string(fake_outputs_count + 1) + " as it is smaller: use a higher ring size");
+          bool own_found = false;
+          existing_ring_found = true;
+          for (const auto &out: ring)
+          {
+            MINFO("Ring has output " << out);
+            if (out < num_outs)
+            {
+              MINFO("Using it");
+              req.outputs.push_back({amount, out});
+              ++num_found;
+              seen_indices.emplace(out);
+              if (out == td.m_global_output_index)
+              {
+                MINFO("This is the real output");
+                own_found = true;
+              }
+            }
+            else
+            {
+              MINFO("Ignoring output " << out << ", too recent");
+            }
+          }
+          THROW_WALLET_EXCEPTION_IF(!own_found, error::wallet_internal_error,
+              "Known ring does not include the spent output: " + std::to_string(td.m_global_output_index));
+        }
+      }
+
+      if (num_outs <= requested_outputs_count && !existing_ring_found)
       {
         for (uint64_t i = 0; i < num_outs; i++)
-          req.outputs.push_back({amount, i});
+          req.outputs.push_back({amount, output_offset + i});
         // duplicate to make up shortfall: this will be caught after the RPC call,
         // so we can also output the amounts for which we can't reach the required
         // mixin after checking the actual unlockedness
         for (uint64_t i = num_outs; i < requested_outputs_count; ++i)
-          req.outputs.push_back({amount, num_outs - 1});
+          req.outputs.push_back({amount, output_offset + num_outs - 1});
       }
       else
       {
         // start with real one
-        uint64_t num_found = 1;
-        seen_indices.emplace(td.m_global_output_index);
-        req.outputs.push_back({amount, td.m_global_output_index});
-        LOG_PRINT_L1("Selecting real output: " << td.m_global_output_index << " for " << print_money(amount));
+        if (num_found == 0)
+        {
+          num_found = 1;
+          seen_indices.emplace(td.m_global_output_index);
+          req.outputs.push_back({amount, td.m_global_output_index});
+          LOG_PRINT_L1("Selecting real output: " << td.m_global_output_index << " for " << print_money(amount));
+        }
 
         // while we still need more mixins
         while (num_found < requested_outputs_count)
```

Comment:
I don't understand that, please elaborate.

Reply:
Maybe my understanding of the ring construction code is incorrect. I'll look into the code more closely.

Reply:
OK, after carefully reading the code, I still think there's an issue. AIUI, the function `get_outs` works as follows:

- `req.outputs` is filled by `selected_transfers.size() * requested_outputs_count` entries where the number `requested_outputs_count` is 150% of the ring size plus some constant (50).

- After the `/get_outs.bin` query to the daemon, the wallet repeats randomly picking outputs from `daemon_resp.outs` in the range `[i*requested_outputs_count, (i+1)*requested_outputs_count)` for the `i`-th input until `outs[i].size()` equals `fake_outputs_count+1`.

The problem lies in the latter: when the key image has a known ring, `req.outputs` gets initialized with those known ring members. Then, in the subsequent `while` loop from L5845, `req.outputs` is further populated using the random selection algorithm based on triangular distribution, until the number of requested outputs for each input reaches `requested_outputs_count`. Finally, the ring members to be used in the resulting tx is selected randomly from the daemon response and gets written to `outs[i]`. Because of this last random selection step, the resulting ring will likely fail to reproduce the known ring.

I confirmed this behavior by tweaking the number `requested_outputs_count`: when it's set to `fake_outputs_count+1`, the known ring gets successfully reproduced. As `requested_outputs_count` grows further away from `fake_outputs_count+1`, the probability of being able to reproduce the known ring declines.

Reply:
I understand now, and I see it's a real bug, thanks.

Reply:
Fixed

Reply:
ffs, not fixed, I made the change on the non-master branch, porting...

Reply:
Fixed